### DISCUSSION TITLE
Resize confusion

### DISCUSSION THREAD

**Discussion by PhillipOConnor** [fltk-rs/fltk-rs#1309] [Q&A] [answered]
I created a data entry application starting from [7guis-fltk-rs crud](https://github.com/tdryer/7guis-fltk-rs/blob/master/src/crud.rs) and found that the window seemed to randomly fail to draw any widgets in the window; the window was all background. I removed large chunks of code and found that a simple button with no callback and no other code did the same thing.

I noticed that the problem was triggered by starting the app from a monitor with different scaling from the monitor on which the app window opened. I then found that when the window was initially correct, any resize would cause it to go blank. The resize could be triggered by dragging the window to the other monitor or by ctrl+ for example.

I started adding in fltk-anchor to control widget positions and found that this had an interesting interaction with the vanishing widgits. On resize the top left anchored active widgets were kept, the others disappeared.

The behaviour is related to the method used to build the window size in the example I started from.  The widgets are added with positions relative to earlier widgets and the window size is set last using the positions of the extreme widgets.

Looking at the code of fltk-anchor it is obvious that the window size has to be known first. What is less obvious to me is that the same problem relates to widgets disappearing on window resize.

The other issue I found with fltk-anchor is that disabled widgets are not anchored.

The following code can be used to display these behaviours, uncomment the first  "win.set_size(W_W, W_H);" to get better behaviour, though disabled widgets are still not correct.

What is the correct way to handle the disabled buttons? Put them in a frame and anchor the frame rather than the button?

```rust
use fltk::{prelude::*, *};
use fltk_anchor::{Anchor, Anchored};

const PADDING: i32 = 8;
const B_W: i32 = 200;
const B_H: i32 = 40;
const C1: i32 = PADDING;
const C2: i32 = C1 + B_W + PADDING;
const C3: i32 = C2 + B_W + PADDING;
const W_W: i32 = C3 + B_W + PADDING;
const R1: i32 = PADDING;
const R2: i32 = R1 + B_H + PADDING;
const R3: i32 = R2 + B_H + PADDING;
const W_H: i32 = R3 + B_H + PADDING;

fn main() {
    let a = app::App::default();
    let mut win = window::Window::default();

    // win.set_size(W_W, W_H);

    let mut b_e1 = button::Button::new(C1, R1, B_W, B_H, "Enable Left Top")
        .with_anchor(Anchor::Left | Anchor::Top);
    let mut b_e2 = button::Button::new(C2, R1, B_W, B_H, "Enable Left-Right Top")
        .with_anchor(Anchor::Left | Anchor::Right | Anchor::Top);
    let mut b_e3 = button::Button::new(C3, R1, B_W, B_H, "Enable Right Top")
        .with_anchor(Anchor::Right | Anchor::Top);
    let mut b_d1 = button::Button::new(C1, R2, B_W, B_H, "Disable Left Top")
        .with_anchor(Anchor::Left | Anchor::Top);
    let mut b_d2 = button::Button::new(C2, R2, B_W, B_H, "Disable Left-Right Top")
        .with_anchor(Anchor::Left | Anchor::Right | Anchor::Top);
    let mut b_d3 = button::Button::new(C3, R2, B_W, B_H, "Disable Right Top")
        .with_anchor(Anchor::Right | Anchor::Top);
    button::Button::new(C1, R3, B_W, B_H, "No anchor 1");
    button::Button::new(C2, R3, B_W, B_H, "No anchor 2");
    button::Button::new(C3, R3, B_W, B_H, "No anchor 3");

    let mut b_d1_c = b_d1.clone();
    let mut b_d2_c = b_d2.clone();
    let mut b_d3_c = b_d3.clone();

    b_d1.deactivate();
    b_d2.deactivate();
    b_d3.deactivate();
    
    b_e1.set_callback(move |_| b_d1_c.activate());
    b_e2.set_callback(move |_| b_d2_c.activate());
    b_e3.set_callback(move |_| b_d3_c.activate());
    b_d1.set_callback(|b| b.deactivate());
    b_d2.set_callback(|b| b.deactivate());
    b_d3.set_callback(|b| b.deactivate());

    win.set_size(W_W, W_H);

    win.end();

    win.make_resizable(true);

    win.show();

    a.run().unwrap();
}
```

**Comment by MoAlyousef** [comment, marked as answer]
The resizing issue seems to be related to this open issue:
https://github.com/fltk-rs/fltk-rs/issues/784

Regarding the fltk-anchor behavior, I’ll have to investigate it further. 

Question, does using `win.resize(win.x(), win.y(), W_W, W_H);` make a difference?

**Reply by MoAlyousef** [reply]
Trying things out with C++:
```cpp
int main() {
    auto win = new Fl_Window(0, 0);
    auto btn = new Fl_Button(100, 100, 80, 30, "Click");
    win->size(400, 300);
    win->end();
    win->resizable(win);
    win->show();
    return Fl::run();
}
```
It seems to be reproducible there as well. I think it's because FLTK's resizing behavior depends on the initial positions of widgets relative to the window's initial size. Wrapping things in a dynamic group like a Flex seems to subvert the issue. So this could be used as a workaround.

**Reply by PhillipOConnor** [reply]
I checked, the `win.resize(win.x(), win.y(), W_W, W_H);` does not help, the solution is to size the window correctly before adding widgets.

I also confirmed that placing the disabled buttons inside a group and attaching the anchor to the group instead of the buttons makes them behave in the same way as the enabled buttons..
